practice-two/src/components/InputValidate/InputValidate.jsx

```diff
@@ -0,0 +1,67 @@
+import './InputValidate.css';
+import PropTypes from 'prop-types';
+import { Validation } from '@helpers';
+import { MESSAGES } from '@constants';
+
+const InputValidate = (props) => {
+  const {
+    type,
+    value,
+    errorMessage,
+    genderType,
+    onChange,
+    errors,
+    setErrors,
+    ...inputProps
+  } = props;
+
+  const handleBlur = (e) => {
+    const { name, value } = e.target;
+    const errorMessage = Validation(name, value);
+    setErrors({
+      ...errors,
+      [name]: errorMessage,
+    });
+  };
+
+  return (
+    <div className='input__group'>
+      <input
+        id={genderType}
+        type={type}
+        value={value}
+        className={type === 'radio' ? 'input__choose' : 'input__field'}
+        onChange={onChange}
+        defaultChecked={value === 'male'}
+        onBlur={handleBlur}
+        {...inputProps}
+      />
+      {type === 'radio' && (
+        <label className='input__option' htmlFor={genderType}>
+          {genderType}
+        </label>
+      )}
+      {errorMessage && <span className='error__message'>{errorMessage}</span>}
+    </div>
+  );
+};
+
+InputValidate.propTypes = {
+  type: PropTypes.oneOf(['text', 'radio']),
+  placeholder: PropTypes.string,
+  errorMessage: PropTypes.string,
+  value: PropTypes.string,
+  name: PropTypes.string,
+  genderType: PropTypes.string,
+};
+
+InputValidate.defaultProps = {
+  type: 'text',
+  placeholder: '',
+  errorMessage: MESSAGES.FORM.USERNAME,
+  value: '',
+  name: '',
+  genderType: '',
+};
+
+export default InputValidate;
```

Comment:
- I will take the following specific actions:
  - Open the form
  - Click submit button
  - Click submit button again to check rerender  

- I check it by 2 ways:
  - **Use React-dev-tool**:
    - When I open form, this is the first time InputValidate component rerendered, so it take a lot of times  
    ![image](https://github.com/datdragon777/React-Training/assets/74806826/6a5a20db-df4d-4609-a3df-7bcccc32ba1f)  
    ![image](https://github.com/datdragon777/React-Training/assets/74806826/d78b470a-ad6a-46cc-8b09-7becd2281112)
    - After that, I click the submit button to check validation
    ![image](https://github.com/datdragon777/React-Training/assets/74806826/cad1fe0e-7224-4587-a28f-acb3fdbbc2ef)
    - Finally, I continue clicking submit button, it still rerender a InputValidate although the value unchange     
    ![image](https://github.com/datdragon777/React-Training/assets/74806826/5a96bec1-a47b-4a46-b58a-d8c8a4f7b6f8)

  - **Use console.log**:
    - I set a console.log in the middle of component
    ![image](https://github.com/datdragon777/React-Training/assets/74806826/a8e5818f-092e-45b9-9c69-5ce9535c324e)
    - I open the form in app and in console window, it will show the log that I write   
    ![image](https://github.com/datdragon777/React-Training/assets/74806826/64843807-4846-4ea0-9a77-1d319601aade)
    - I clear the old log and press the submit button, it will log like photo above
    - But I press submit button again, I still print the log even if the value in each input unchange
    ![image](https://github.com/datdragon777/React-Training/assets/74806826/124d8f8c-b2eb-4a55-aef0-623f34e3f21f)
    
- Even if I use Reac.memo for child component (InputValidate), the way I fix this:
  - I use useCallback and previous state
  - useCallback is used when memo is not effect, and previous state is used for storing the previous result, so component won't rerender unnecessary
  - The third time is descreased than the way before
  ![image](https://github.com/datdragon777/React-Training/assets/74806826/be788d71-deba-4a87-8dbc-3ed1ae7ee55e)